rfcs/2021-11-12-9811-vrl-vm.md

```diff
@@ -0,0 +1,347 @@
+tree is boxed and stored in disparate regions of memory. As a result walking
+the tree means that the CPU caches must be constantly swapped.
+
+## Proposal
+
+### Implementation
+
+Instead we can create an enum VM to store the execution of the Vrl program.
+
+The enum is essentially a big enum of instructions:
+
+```rust
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum OpCode {
+    Return,
+    Constant,
+    Negate,
+    Add,
+    Subtract,
+    Multiply,
+    Divide,
+    Print,
+    Not,
+    Greater,
+    GreaterEqual,
+    Less,
+    LessEqual,
+    NotEqual,
+    Equal,
+    Pop,
+    JumpIfFalse,
+    Jump,
+    SetPath,
+    GetPath,
+    Call,
+    ...
+}
+
+enum Instruction {
+    Opcode(Opcode),
+    Literal(LiteralIndex),
+}
+
+pub struct LiteralIndex(usize);
+```
+
+The Vm is a struct comprising of the following fields:
+
+```rust
+#[derive(Clone, Debug, Default)]
+pub struct Vm {
+    instructions: Vec<Instruction>,
+    constants: Vec<Literal>,
+    targets: Vec<Variable>,
+    stack: Vec<Value>,
+    parameter_stack: Vec<Option<Value>>,
+    error: Option<Error>,
+    instruction_pointer: usize,
+}
+```
+
+- instructions
+
+The instructions field is a `Vec` of `Instruction`. An instruction can be
+either an OpCode or some data for that OpCode. For example, the instructions
+`[.., OpCode(Constant), Literal(12), ..]` when evaluated will load the constant
+stored in the `values` `Vec` that is found in position 12 onto the stack.
+
+- constants
+
+A list of constant values found in the program. Since the bytecode only
+contains integers any actual values must be stored here. This also allows
+literals to be deduped.
+
+- targets
+
+A list of paths used in the program, similar to `values`.
+
+- stack
+
+The Vm is a stack based Vm. Every expression that is evaluated pushes the
+result on the stack. Every operation pulls the values it uses from the stack.
+
+- parameter_stack
+
+For function calls, the parameters need to be evaluated and passed to the
+function. The stack is a `Vec` of parameter name and value. We need an OpCode
+that will copy a value from the main stack to the parameter stack tagged with
+the parameter name. The `FunctionArgs` compiler will dump this OpCode after the
+code to evaluate the given arg is dumped.
+
+- error
+
+If an operation errors, it needs to populate this field. Operations such as
+the error assignment (`result, error = thing()`) will need to check if this
+field has been populated and action accordingly.
+
+- instruction_pointer
+
+The instruction pointer points to the next instruction to evaluate.
+
+#### Calling functions
+
+Calling functions in the stdlib will be a case of evaluating each parameter
+with the results pushed onto the parameter stack.
+
+Since VRL allows for named parameters, parameters need to be pushed on to the
+parameter stack _in the order that they are specified in the `Function`
+implementation_.
```

Comment:
Is this correct? Or, is this the complete story?

I'm still unsure how this solves the problem of VRL programs allowing function calls to mix-and-match named and positional arguments _and_ allowing named arguments to be anywhere in the function call, i.e. there is no pre-defined order for named parameters.

Reply:
The compiler would need to re-order the parameters at the call site to match the function definition. IOW if we have a function defined as `foo(a, b, c)` and the program calls `foo(1, c=2, b=3)`, the compiler would need to reorder it to `foo(1, 3, 2)`, and the called function can then pull the values off the stack in a predictable order.

Reply:
Got it, that makes perfect sense 👍